src/core/cfg.cpp

```diff
@@ -2274,105 +2222,197 @@ class CFGVisitor : public ASTVisitor {
-            break_dest = cfg->addDeferredBlock();
-            break_dest->info = "with_break";
 
-            pushLoopRegion(continue_dest, break_dest);
-        }
+        // push continuations
+        CFGBlock* finally_block = cfg->addDeferredBlock();
+        finally_block->info = "with_finally";
+        pushFinallyContinuation(finally_block, whyname);
 
-        CFGBlock* return_dest = cfg->addDeferredBlock();
-        return_dest->info = "with_return";
-        pushReturnRegion(return_dest);
+        CFGBlock* exc_block = cfg->addDeferredBlock();
+        exc_block->info = "with_exc";
+        exc_handlers.push_back({ exc_block, exc_type_name, exc_value_name, exc_traceback_name });
 
         for (int i = 0; i < node->body.size(); i++) {
             node->body[i]->accept(this);
         }
 
-        popRegion(); // for the retrun
-
-        AST_Call* exit_call = makeCall(makeName(exitname, AST_TYPE::Load, node->lineno));
-        exit_call->args.push_back(makeName(nonename, AST_TYPE::Load, node->lineno));
-        exit_call->args.push_back(makeName(nonename, AST_TYPE::Load, node->lineno));
-        exit_call->args.push_back(makeName(nonename, AST_TYPE::Load, node->lineno));
-        push_back(makeExpr(exit_call));
+        exc_handlers.pop_back();
+        int finally_did_why = continuations.back().did_why;
+        popContinuation();
 
-        CFGBlock* orig_ending_block = curblock;
+        if (curblock) {
+            // The try-suite finished as normal; jump to the finally block.
+            pushAssign(whyname, makeNum(Why::FALLTHROUGH));
+            pushJump(finally_block);
+        }
+
+        // This tells us whether our exit_block will have multiple incoming edges (one from exc_block, one from
+        // finally_block) or not (if exc_block and/or finally_block is unneeded, or if finally_block doesn't jump to
+        // exit_block because it's never entered except via continue, break, and/or return). If exit_block has multiple
+        // incoming edges, then anybody *branching* into it (rather than jumping directly) needs to do so through a
+        // "trampoline" block to avoid critical edges.
+        // bool exit_multi_incoming = exc_block->predecessors.size() && (finally_did_why & (1 << Why::FALLTHROUGH));
+        bool exit_multi_incoming = true; // FIXME FIXME FIXME(rntz)
+
+        // The exception-handling block
+        if (exc_block->predecessors.size() == 0) {
+            // TODO(rntz): test for this case
+            delete exc_block;
+        } else {
+            cfg->placeBlock(exc_block);
+            curblock = exc_block;
+
+            // call the context-manager's exit method
+            InternedString suppressname = nodeName(node, "suppress");
+            pushAssign(suppressname, makeCall(makeLoad(exitname, node), makeLoad(exc_type_name, node),
+                                              makeLoad(exc_value_name, node), makeLoad(exc_traceback_name, node)));
```

Comment:
I think you may need to call `callNonzero()` on the result of makeCall -- it looks like `__exit__` is allowed to return a non-boolean type, and at some point we have to coerce that to an unboxed bool, which is what the NONZERO ast node is for.